tests/test_rp_exec.py

```diff
@@ -50,97 +51,165 @@ def get_rp_decorator():
+        tmgr  = rp.TaskManager(session=session)
+
+        pilot = pmgr.submit_pilots(pd)
+        task  = tmgr.submit_tasks(td)
+
+        tmgr.add_pilots(pilot)
+        tmgr.wait_tasks(uids=[task.uid])
+
+        assert task.exit_code == 0
+
+    assert session.closed
 
 
+# ------------------------------------------------------------------------------
+#
 # Note: radical.pilot.Session creation causes several deprecation warnings.
 # Ref https://github.com/radical-cybertools/radical.pilot/issues/2185
+#
 @pytest.mark.filterwarnings('ignore::DeprecationWarning')
 @with_radical_only
-def test_rp_basic_task(rp_config):
-    rp = rp_config['rp']
+def test_rp_scalems():
+
+    import radical.pilot as rp
 
-    # Note: Session creation will fail with a FileNotFound error unless venv is explicitly `activate`d.
-    # TODO: Figure out what `activate` does that `rp-venv/bin/python` doesn't do.
     with rp.Session() as session:
-        # Based on `radical.pilot/examples/config.json`
-        # TODO: Does the Session have a default spec for 'local.localhost'? Can/should we reference it?
-        # See also https://github.com/radical-cybertools/radical.pilot/issues/2181
-        resource = 'local.localhost'
-        resource_config = {resource: {}}
-        if resource in rp_config['config']:
-            resource_config[resource].update(rp_config.config[resource])
-        resource_config[resource].update({
-            'project': None,
-            'queue': None,
-            'schema': None,
-            'cores': 1,
-            'gpus': 0
-        })
-
-        pilot_description = dict(resource=resource,
-                                 runtime=30,
-                                 exit_on_error=True,
-                                 project=resource_config[resource]['project'],
-                                 queue=resource_config[resource]['queue'],
-                                 cores=resource_config[resource]['cores'],
-                                 gpus=resource_config[resource]['gpus'])
-
-        task_description = {'executable': '/bin/date',
-                            'cpu_processes': 1,
-                            }
-
-        pmgr = rp.PilotManager(session=session)
-        umgr = rp.UnitManager(session=session)
-        pilot = pmgr.submit_pilots(rp.ComputePilotDescription(pilot_description))
-        task = umgr.submit_units(rp.ComputeUnitDescription(task_description))
-
-        umgr.add_pilots(pilot)
-        umgr.wait_units()
 
-        assert task.exit_code == 0
+        # define a pilot and launch it
+        pd    = rp.PilotDescription(
+                   {'resource': 'local.localhost',
+                    'cores'   : 4,
+                    'gpus'    : 0})
+
+        pmgr  = rp.PilotManager(session=session)
+        tmgr  = rp.TaskManager(session=session)
+        pilot = pmgr.submit_pilots(pd)
+        tmgr.add_pilots(pilot)
+
+        # TODO: How can we recover successful components of previous failed Sessions?
```

Comment:
With `components` you mean `tasks`, so components of the ScaleMS workflow?  If so, their state / results need to be recorded out of band, RP will not track them beyond the lifetime of a session.

Reply:
I'll update to the following:
```
        # TODO: How can we recover successful workflow stages from previous failed Sessions?
        #    The client needs to note the sandbox locations from runs. SCALEMS can
        #    then manage / maintain state tracking or state discovery to optimize workflow recovery.
        #    Resumed workflows can make reference to sandboxes from previous sessions (RCT work in progress).
```
Can you remind me which part of this is an RCT work in progress? Was it something to do with the robust/proper/easy way to make reference to prior sandboxes?

Reply:
RP works toward allowing these kinds of sandbox references for data staging URLs:
```
    sandbox://task.0001/foo/bar/data.bin
    sandbox://pilot.0000/file.data
    sandbox://rp.session...../foo.dat
    sandbox://xsede.stampede2/some/dir/some/file.txt
    sandbox://client/foo.bar
```

Reply:
Would "sandbox" be the UUID of the sandbox? Or the key word "sandbox" with the UUID following the `://`?

What is the current status? We would need to construct `file://$OLD_SANDBOX/task.0001/foo/bar/data.bin`?

Reply:
`sandbox://` would be a static  schema.  Maybe `rp://` would be better actually?  Naming, as hard as ever... :-P
In that URL schema, the `hostname` element would be interpreted as UID or a task, session, pilot etc., and the URL would reference the sandbox of the thus identified entity.

> What is the current status? We would need to construct file://$OLD_SANDBOX/task.0001/foo/bar/data.bin?

basically yes.  Note that a completed task should have a `sandbox` attribute set which points to the used sandbox:
```py
        task_dict = tasks[0].as_dict()
        report.plain("task workdir : %s\n" % task_dict['task_sandbox'])
        report.plain("pilot id     : %s\n" % task_dict['pilot'])
        report.plain("exit code    : %s\n" % task_dict['exit_code'])
        report.plain("stdout       : %s\n" % task_dict['stdout'])
```
would result in,  for example:
```sh
task workdir : file://localhost/home/merzky/radical.pilot.sandbox/rp.session.rivendell.merzky.018683.0018/pilot.0000/task.000000/
pilot id     : pilot.0000
exit code    : 0
stdout       : Thu Feb 25 17:24:28 CET 2021
```

Reply:
Excellent. Thanks.